### PR TITLE
fpstate-related cleanups and speedups

### PR DESCRIPTION
**Pull request by bartoldeman**
Relates to #572.

I haven't submitted any of my proposed changes to dpmi.c yet, that needs some more testing. There is also still the issue that SSE state is not saved/restored on i386.

### CODE REVIEW

File: src/dosext/dpmi/dpmi.c [hunk 1]
```diff
@@ -1158,7 +1157,7 @@ static void dpmi_switch_sa(int sig, siginfo_t *inf, void *uc)
   ucontext_t *uct = uc;
   sigcontext_t *scp = &uct->uc_mcontext;
 
-  emu_stack_frame.fpregs = &emu_fpstate;
+  emu_stack_frame.fpregs = aligned_alloc(16, sizeof(*__fpstate));
```
Comment (stsp) [inline src/dosext/dpmi/dpmi.c]: Maybe to dpmi_setup() and never free()?
After all emu_stack_frame is static.

**Comment by stsp** [comment]
> Don't save/restore FPU state for KVM

Do you mean _dosemu_ FPU state?
Or guest's state, or both?
Looking into the patch, it seems to be
dosemu state, so what do we do with
the guest state in kvm?

**Comment by stsp** [comment]
OK, it looks you don't save both because
KVM has it separately.

**Comment by stsp** [comment]
But do we at least have the pointer
to KVMs FPU state? Can it be easily
saved if we, for example, want the
ability to write save snapshots?

**Comment by bartoldeman** [comment]
For KVM, you can get and set FPU state via the ioctls 4.22 KVM_GET_FPU and 4.23 KVM_SET_FPU, see
https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt
so it is straightforward to obtain on demand.

If I allocate emu_stack_frame.fpregs in dpmi_setup() it should be inside
`if (config.cpu_vm_dpmi == CPUVM_NATIVE)`, as emu_stack_frame is really only used for that case.

The other "no-copy" approach (as discussed in #572) is to never allocate it, return to dosemu with DPMI fpu state and then simply call `fesetenv(&dosemu_env);`

**Comment by stsp** [comment]
Ok, thanks.
I think the still unsolved problem is the FPU
reset on sighandler entry, right?

**Comment by bartoldeman** [comment]
Yes, I will test to confirm but it looks like it does (as well as clearing DF, RF, and TF in eflags):
https://github.com/torvalds/linux/blame/35277995e17919ab838beae765f440674e8576eb/arch/x86/kernel/signal.c#L743
the function fpu__clear() is called when
1. entering a signal handler
2. if returning from a signal handler with NULL in the fpstate pointer
3. if returning from a signal handler with problematic data in the fpstate (an error was found)
4. when flushing a thread

**Comment by stsp** [comment]
Ok, then nothing to worry about.
Thanks for taking a look.

**Comment by stsp** [comment]
> Yes, I will test to confirm but it looks like it does (as well as clearing DF, RF, and TF in eflags):

But forgets to clear AC and NT IIRC?

**Pull request merged**